File: Tree/333.Largest-BST-Subtree.java
/*
https://leetcode.com/problems/largest-bst-subtree/
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class InfoNode {
    int max; // max value in subtree rooted at node
    int min; // min value in subtree rooted at node
    boolean isBST; // is the subtree rooted at node a BST
    int currMaxSize; // the current maximum size of a BST in tree so far
    
    public InfoNode(int max, int min, boolean isBST, int currMaxSize) {
        this.max = max;
        this.min = min;
        this.isBST = isBST;
        this.currMaxSize = currMaxSize;
    }
}


class Solution {    
    public int largestBSTSubtree(TreeNode root) {
        InfoNode res = helper(root);
        return res.currMaxSize;
    }
    
    private InfoNode helper(TreeNode root) {
        if(root == null) {
            return new InfoNode(Integer.MIN_VALUE, Integer.MAX_VALUE, true, 0);
        }
        
        InfoNode leftInfo = helper(root.left);
        InfoNode rightInfo = helper(root.right);
        
        int maxSubtreeAtRoot = Math.max(root.val, Math.max(leftInfo.max, rightInfo.max));
        int minSubtreeAtRoot = Math.min(root.val, Math.min(leftInfo.min, rightInfo.min));
        
        if(leftInfo.isBST && rightInfo.isBST && leftInfo.max < root.val && root.val < rightInfo.min) { // potential for subtree rooted at root to become a BST
            
            int currMaxSizeAtRoot = leftInfo.currMaxSize + rightInfo.currMaxSize + 1;
            return new InfoNode(maxSubtreeAtRoot, minSubtreeAtRoot, true, currMaxSizeAtRoot);
            
        } else {
            
            int currMaxSizeAtRoot = Math.max(leftInfo.currMaxSize, rightInfo.currMaxSize);
            return new InfoNode(maxSubtreeAtRoot, minSubtreeAtRoot, false, currMaxSizeAtRoot);
        }
        
    }
    
}

File: Tree/333.Largest-BST-Subtree.py
# https://leetcode.com/problems/largest-bst-subtree/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

MAX = 10001
MIN = -10001

class NodeInfo:
    
    def __init__(self, max, min, isBST, currMaxSize):
        self.max = max  # max value in subtree 
        self.min = min  # min value in subtree
        self.isBST = isBST # whether subtree is a BST
        self.currMaxSize = currMaxSize  # size of current max BST (from root downwards, isn't necessarily rooted at current node. In other words, somewhere in the subtree rooted at root, there happens to be a maximum size BST (so far))
        
        
class Solution:
    def largestBSTSubtree(self, root: TreeNode) -> int:
        node_info = self.helper(root)
        return node_info.currMaxSize
    
    def helper(self, root: TreeNode) -> NodeInfo:
        if root is None:
            return NodeInfo(MIN, MAX, True, 0)
        
        info_left = self.helper(root.left)
        info_right = self.helper(root.right)
        
        currMax = max(root.val, max(info_left.max, info_right.max))
        currMin = min(root.val, min(info_left.min, info_right.min))
        
        if info_left.isBST and info_right.isBST and info_left.max < root.val and root.val < info_right.min:
            currMaxSize = info_left.currMaxSize + info_right.currMaxSize + 1
            return NodeInfo(currMax, currMin, True, currMaxSize)
        else:
            currMaxSize = max(info_left.currMaxSize, info_right.currMaxSize)
            return NodeInfo(currMax, currMin, False, currMaxSize)
        
        
